File: proyecto/cppcalc/calculator.cpp
#include "calculator.h"
#include "parser.h"
#include <sstream>


Calculator::Calculator() :
        memory(0), variableMemory(), files() { }

int Calculator::eval(string expr) {

    Parser *parser = new Parser(new istringstream(expr));
    AST *tree = parser->parse();
    int result = tree->evaluate();

    delete tree;
    delete parser;
    return result;
}

void Calculator::store(int val) {
    memory = val;
}

int Calculator::recall() {
    return memory;
}

int Calculator::plus(int n) {
    memory += n;
    return memory;
}

int Calculator::minus(int n) {
    memory -= n;
    return memory;
}

int Calculator::clear() {
    memory = 0;
}

void Calculator::storevar(string name, int val){
    variableMemory[name] = val;
}

int Calculator::recallvar(string name){
    if(variableMemory[name])
        return variableMemory[name];
    else
        return 0;
}


void Calculator::setPrefEnv(bool set){
    preferEnv = set;
}

bool Calculator::getPreferEnv(){
    return preferEnv;
}

void Calculator::setAssignPerformed(bool assignPerformed){
    Calculator::assignPerformed = assignPerformed;
}

bool Calculator::isAssignPerformed(){
    return assignPerformed;
}

void Calculator::addFile(string filename)
{
    files.push_back(filename);
}

deque<string> Calculator::getQueue() const
{
    return files;
}


int Calculator::getFileNumber()
{
    return files.size();
}

File: proyecto/rubycalc/lib/calculator.rb
require 'parser'
require 'ast'

class Calculator
  attr_accessor :memory, :assigned, :variables, :envprefered, :assignlog

  def initialize
    @memory = 0
    @variables = Hash.new
    @assigned = false
    @envprefered = false
    @assignlog = Array.new
  end

  def eval(expr)
    expr = expr.gsub(/\s+/, " ")
    parser = Parser.new(StringIO.new(expr))
    ast = parser.parse
    ast.evaluate
  end

  def clear
    @memory = 0
  end

  def storevar_(name, value, set)
    if set
      @assigned = true
      @assignlog.push(name + " <- " + value.to_s)
    end
    @variables.store(name.chomp, value)
  end

  def storevar(name, value)
    storevar_(name, value, true)
  end

  def storevar!(name, value)
    storevar_(name, value, false)
  end

  def getvar(name)
    @variables.fetch(name.chomp, 0)
  end

  def store(num)
    @memory = num
  end

  def recall
    @memory
  end

  def mem_minus(num)
    @memory -= num
  end

  def mem_plus(num)
    @memory += num
  end
end
